narciuch/assignment4/src/IslandFinder.java

```diff
@@ -69,43 +67,40 @@ public int getIslands(int nrRows, int nrColumns, boolean[][] lands) {
-   * array that it keeps in the islands array.
-   *
+   * Merges the values of islands, i.e the integer sets, that have keys specified by the islandKeys,
+   * into one set that it keeps in islands, i.e. merges multiple islands into one.
+   * <p>
    * <p>Returns a new copy of islands after this merge.
    *
-   * @param islands       a list of integer lists, where each inner integer list represents an island
-   * @param islandIndexes the indices of those inner arrays in 'islands' that are to be merged
+   * @param islands    a mapping from a unique integer key to a set of integers,
+   *                   where each integer set represents an island
+   * @param islandKeys the keys of those sets in 'islands' that are to be merged
    * @return a new copy of islands which has merged the islands as specified and not changed the
-   * other islands within the 'islands' nested list
+   * other islands within 'islands'
    */
-  private List<List<Integer>> mergeIslands(List<List<Integer>> islands, List<Integer> islandIndexes) {
-    List<Integer> mergeInto = islands.get(islandIndexes.get(0));
-    for (int mergeIslandIndex = 1; mergeIslandIndex < islandIndexes.size(); mergeIslandIndex++) {
-      int islandNr = islandIndexes.get(mergeIslandIndex);
-      for (int tileNr : islands.get(islandNr)) {
-        mergeInto.add(tileNr);
-      }
-      islands.remove(islandNr);
-      islands.add(islandNr, new ArrayList<>());
+  private Map<Integer, Set<Integer>> mergeIslands(Map<Integer, Set<Integer>> islands, List<Integer> islandKeys) {
+    Set<Integer> mergeInto = islands.get(islandKeys.get(0));
+    for (int mergeIslandIndex = 1; mergeIslandIndex < islandKeys.size(); mergeIslandIndex++) {
+      int islandKey = islandKeys.get(mergeIslandIndex);
+      mergeInto.addAll(islands.get(islandKey));
+      islands.remove(islandKey);
     }
-    return islands.stream().filter(island -> island.size() != 0).collect(Collectors.toList());
+    return islands;
   }
 
 
   /**
-   * We assume each tilePosition is in the islands list at most once.
+   * We assume each tilePosition is in islands at most once.
    *
    * @param islands
    * @param tilePosition
-   * @return index of the island containing the tilePosition value if it exists in the list
-   * of islands. If the tilePosition is not in the list of islands this returns -1.
+   * @return key of the island containing the tilePosition value if it exists in 'islands'.
+   * If the tilePosition is not in 'islands' then -1 is returned.
    */
-  private int getIslandIndex(List<List<Integer>> islands, int tilePosition) {
-    int totalIslandNr = islands.size();
-    for (int islandNr = 0; islandNr < totalIslandNr; islandNr++) {
-      if (islands.get(islandNr).contains(tilePosition)) {
-        return islandNr;
+  private int getIslandKey(Map<Integer, Set<Integer>> islands, int tilePosition) {
+    for (int key : islands.keySet()) {
```

Comment:
Optional: if you used `Map.entrySet()` for iteration, you could avoid looking key up with `get(key)` below.